src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs

```diff
@@ -302,174 +305,99 @@ public class OpenIdConnectAuthenticationHandler : AuthenticationHandler<OpenIdCo
+                    ticket = ValidateToken(message.IdToken, message, properties, validationParameters, out jwt);
 
-                if (message.Code != null)
-                {
-                    Logger.LogDebug(Resources.OIDCH_0014_AuthorizationCodeReceived, message.Code);
-                    if (ticket == null)
+                    var securityTokenValidatedNotification = await RunSecurityTokenValidatedNotificationAsync(message, ticket);
+                    if (securityTokenValidatedNotification.HandledResponse)
                     {
-                        ticket = new AuthenticationTicket(properties, Options.AuthenticationScheme);
+                        return securityTokenValidatedNotification.AuthenticationTicket;
                     }
-
-                    var authorizationCodeReceivedNotification = new AuthorizationCodeReceivedNotification(Context, Options)
-                    {
-                        AuthenticationTicket = ticket,
-                        Code = message.Code,
-                        JwtSecurityToken = jwt,
-                        ProtocolMessage = message,
-                        RedirectUri = ticket.Properties.Items.ContainsKey(OpenIdConnectAuthenticationDefaults.RedirectUriForCodePropertiesKey) ?
-                                      ticket.Properties.Items[OpenIdConnectAuthenticationDefaults.RedirectUriForCodePropertiesKey] : string.Empty,
-                    };
-
-                    await Options.Notifications.AuthorizationCodeReceived(authorizationCodeReceivedNotification);
-                    if (authorizationCodeReceivedNotification.HandledResponse)
+                    else if (securityTokenValidatedNotification.Skipped)
                     {
-                        Logger.LogVerbose(Resources.OIDCH_0015_AuthorizationCodeReceivedNotificationHandledResponse);
-                        return authorizationCodeReceivedNotification.AuthenticationTicket;
+                        return null;
                     }
 
-                    if (authorizationCodeReceivedNotification.Skipped)
+                    ValidateOpenIdConnectProtocolAsync(jwt, message);
+
+                    if (message.Code != null)
                     {
-                        Logger.LogVerbose(Resources.OIDCH_0016_AuthorizationCodeReceivedNotificationSkipped);
-                        return null;
+                        var authorizationCodeReceivedNotification = await RunAuthorizationCodeReceivedNotificationAsync(message, properties, ticket, jwt);
+                        if (authorizationCodeReceivedNotification.HandledResponse)
+                        {
+                            return authorizationCodeReceivedNotification.AuthenticationTicket;
+                        }
+                        else if (authorizationCodeReceivedNotification.Skipped)
+                        {
+                            return null;
+                        }
                     }
                 }
```

Comment:
It shouldn't really happen but having a debug log makes sense. I will add the else scenario.